killbill-integration-tests/plugins/plugin_base.rb

```diff
@@ -1,36 +1,124 @@
+      prepare_teardown_sequence(plugin_key) do |seq|
+        run_plugin_sequence("stop", plugin_key, nil, [], seq)
+      end
       teardown_base
     end
 
+    def set_configuration(plugin_name, plugin_config)
+      KillBillClient::Model::Tenant.upload_tenant_plugin_config(plugin_name,
+                                                                plugin_config,
+                                                                @user,
+                                                                nil,
+                                                                nil,
+                                                                @options)
+      # make sure to settle
+      sleep 5
+    end
 
     private
 
+    def prepare_teardown_sequence(plugin_key)
+      seq = %i[stop_plugin uninstall_plugin]
+
+      plugin_state_file = plugin_key + '_temp_state.yml'
+      if File.exist?(plugin_state_file)
+        plugin_state = YAML.load_file(plugin_state_file)
+        seq -= [:stop_plugin] unless plugin_state[:stop_plugin]
+        seq -= [:uninstall_plugin] unless plugin_state[:uninstall_plugin]
+      end
+
+      yield seq unless seq.empty?
+    end
+
+    def prepare_setup_sequence(plugin_key, plugin_version)
+      seq = %i[install_plugin start_plugin]
+      plugins_info = get_plugin_information(plugin_key, plugin_version)
+
+      is_installed = !(plugins_info.nil? || plugins_info.empty?)
+      is_installed_and_running = is_installed && plugins_info.first.state == 'RUNNING'
+      seq -= [:install_plugin] if is_installed
+
+      # store stop sequence for the teardown, since it will run on every test
+      plugin_state_file = plugin_key + '_temp_state.yml'
+      File.delete(plugin_state_file) if File.exist?(plugin_state_file)
+      File.new(plugin_state_file, 'w+')
+
+      plugin_state = YAML.load_file(plugin_state_file) || {}
+      plugin_state[:stop_plugin] = !is_installed_and_running
+      plugin_state[:uninstall_plugin] = !is_installed
+
+      File.open(plugin_state_file, 'r+') do |f|
+        f.write(plugin_state.to_yaml)
+      end
+
+      puts "\e[36m#{plugin_key} is already installed\e[0m" if is_installed && !is_installed_and_running
+      puts "\e[36m#{plugin_key} is already installed and running\e[0m" if is_installed_and_running
+      return plugins_info.first if is_installed_and_running
+      yield seq
+
+      plugins_info = get_plugin_information(plugin_key, plugin_version) unless is_installed
+      plugins_info.first
+    end
+
+    def get_plugin_information(plugin_key, plugin_version)
+      nodes_info = KillBillClient::Model::NodesInfo.nodes_info(@options)
+      return [] if nodes_info.nil?
+      latest_version = Gem::Version.new('0.0.0')
+      has_running_plugin = false
+
+      plugins_info = nodes_info.first.plugins_info || []
+      plugins_info.select! do |plugin|
+        found = true
+        found &= plugin.plugin_key.include?(plugin_key) unless plugin.plugin_key.nil?
```

Comment:
A second email notifications plugin is still installed for me, and I believe this is the cause.

I installed it using the dev mode instructions from https://github.com/killbill/killbill-cloud/tree/master/kpm#dev-mode, i.e.

```
kpm install_java_plugin 'dev:email' --from-source-file=./target/killbill-email-notifications-plugin-0.3.1-SNAPSHOT.jar
```

so a custom key is specified. I think we should match either on the key or the name.